### PR TITLE
Remove ancient register allocation

### PR DESCRIPTION
**Pull request by bnjbvr**
There's still a lot that can be done, but that'll be enough playing for tonight :-)

This removes the old register allocation, and the entire folder `regalloc` that contained related code, like live range analysis, etc. and quite a lot in binemit too (maybe that can be entirely removed too with additional effort?).

### CODE REVIEW

**Review by cfallin — APPROVED** [review]
Today is a good day for deleting code! Thanks a bunch for this!

File: cranelift/codegen/src/context.rs [hunk 1]
```diff
@@ -192,25 +167,11 @@ impl Context {
...
-            }
-            if opt_level == OptLevel::SpeedAndSize {
-                self.shrink_instructions(isa)?;
-            }
-            let result = self.relax_branches(isa);
-
-            log::trace!("Compiled:\n{}", self.func.display(isa));
-            result
-        }
+        let backend = isa.get_mach_backend().expect("only mach backends nowadays");
```
Comment (bjorn3) [inline cranelift/codegen/src/context.rs:170]: Can you add a FIXME to make this non-optional?

Comment (bnjbvr) [reply cranelift/codegen/src/context.rs:170]:
```suggestion
        // FIXME: make this non optional
        let backend = isa.get_mach_backend().expect("only mach backends nowadays");
```

Comment (cfallin) [reply cranelift/codegen/src/context.rs:171]: More generally, I expect that we'll continue to tweak the data structures to remove `Option<>`s on new bits and delete the old bits; the CLIF function representation still has some bits that can go (`jt_offsets` for example?) too.

Comment (bnjbvr) [reply cranelift/codegen/src/context.rs:171]: I'd expect that too, yeah. The fact that so many things were marked `pub` (and not `pub(crate)`) make it so the compiler is not seeing all the dead code out there.

File: cranelift/codegen/src/cursor.rs [hunk 2]
```diff
@@ -759,11 +759,7 @@ impl<'c, 'f> ir::InstInserterBase<'c> for &'c mut EncCursor<'f> {
         &mut self.func.dfg
```
Comment (bjorn3) [inline cranelift/codegen/src/cursor.rs:759]: Can `EncCursor` be removed entirely? It doesn't update encodings anymore.

Comment (bnjbvr) [reply cranelift/codegen/src/cursor.rs:759]: Sounds like yes!

**Pull request merged**